rust/src/dirdiff.rs

```diff
@@ -100,27 +102,34 @@ fn file_content_changed(
 }
 
 fn is_changed(
-    src: &openat::Dir,
-    dest: &openat::Dir,
+    src: &Dir,
+    dest: &Dir,
     path: &str,
-    srcmeta: &openat::Metadata,
-    destmeta: &openat::Metadata,
+    srcmeta: &Metadata,
+    destmeta: &Metadata,
 ) -> Result<bool> {
+    use cap_primitives::fs::read_link_contents;
     if srcmeta.permissions() != destmeta.permissions() {
         return Ok(true);
     }
-    Ok(match srcmeta.simple_type() {
-        openat::SimpleType::File => {
-            if srcmeta.len() != destmeta.len() {
-                true
-            } else {
-                file_content_changed(src, dest, path, srcmeta.len())?
-            }
+    let t = srcmeta.file_type();
+    let r = if t.is_file() {
+        if srcmeta.len() != destmeta.len() {
+            true
+        } else {
+            file_content_changed(src, dest, path, srcmeta.len())
+                .context("Comparing file content")?
         }
-        openat::SimpleType::Symlink => src.read_link(path)? != dest.read_link(path)?,
-        openat::SimpleType::Other => false,
-        openat::SimpleType::Dir => false,
-    })
+    } else if t.is_symlink() {
+        let src_target =
+            read_link_contents(&src.as_filelike_view(), Path::new(path)).context("readlink src")?;
+        let dest_target = read_link_contents(&dest.as_filelike_view(), Path::new(path))
+            .context("readlink dest")?;
```

Comment:
Nice that we're adding error contexts, though would be good to include the actual filename that failed.

Reply:
This is an interesting topic for sure.  Doing so is honestly a bit tedious :smile: so I want to debate/discuss.

I ended up sprinking these contexts around because I was trying to debug where an error was coming from and somehow completely missed that there were *two* loops in `diff_recurse`.

Now I definitely think it's valuable to ensure we have file paths for anything that could be "user facing" - particuarly where *user provided filenames* are in play.

But in this case, we're traversing a filesystem tree that was not explicitly typed by the user.  I have trouble thinking of a case where we'd hit an error on just e.g. one of the files.

(Again, what caused me to sprinkle these contexts was me missing the need to use `.symlink_metadata_optional()` and not `metadata_optional()` in a different section of the code)

So I've chosen to address this by removing all added contexts, for now.  I think we should view this context debate orthogonally to the cap-std port.  We can do more contexts as followup.

Reply:
I can agree we improve the error-cases UX later on.
For reference, I had a similar recursive error context to handle in a previous porting effort, and this is the way I approached it:
https://github.com/coreos/rpm-ostree/blob/fba52302118756fad033bbb5eade44c4b3d99f61/rust/src/composepost.rs#L750-L758

Reply:
I'm cool with removing the contexts for now. :+1: 

Interesting topic indeed! I think that wrong context could be worse than no context (though most contexts here were not completely wrong to be clear!). Also, contexts are not just useful for users. It makes the error users report to us easier to pinpoint. So I wouldn't discard the idea of adding contexts even in deeply nested code. And it's especially in the places you least expect errors where good contexts would be helpful!

To that point, in libostree and rpm-ostree we routinely retroactively sprinkle error prefixing when users/pipelines have already hit an issue. In coreos-installer, we've gotten used to consistently adding contexts (mostly thanks to @bgilbert!) and while it's tedious at times, I can say it does make a difference in diagnosis.